### PR TITLE
[eventhub] Custom Endpoint 

### PR DESCRIPTION
**Pull request by l0lawrence**
# Description

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-python/blob/main/CONTRIBUTING.md##building-and-testing)
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py [hunk 1]
```diff
@@ -268,6 +268,14 @@ def from_connection_string(cls, conn_str, consumer_group, **kwargs):
          Greedy strategy is used by default.
         :paramtype load_balancing_strategy: str or ~azure.eventhub.LoadBalancingStrategy
         :rtype: ~azure.eventhub.EventHubConsumerClient
+        keyword str custom_endpoint_address: The custom endpoint address to use for establishing a connection to
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py:271]:
```suggestion
        :keyword str custom_endpoint_address: The custom endpoint address to use for establishing a connection to
```

Comment (annatisch) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_consumer_client.py:271]: All of the lines trailing this one (272-275) will need to be indented by a single space character. Same applies to lines 277-278.

File: sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py [hunk 2]
```diff
@@ -58,6 +58,22 @@ class EventHubProducerClient(ClientBase):
     :keyword float idle_timeout: Timeout, in seconds, after which this client will close the underlying connection
      if there is no activity. By default the value is None, meaning that the client will not shutdown due to inactivity
      unless initiated by the service.
+    :keyword transport_type: The type of transport protocol that will be used for communicating with
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_producer_client.py:61]: Were these not documented in the PR that added the websocket support?

File: sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_connection.py [hunk 3]
```diff
@@ -90,6 +90,9 @@ def __init__(self, endpoint, **kwargs):
         # type(str, Any) -> None
         parsed_url = urlparse(endpoint)
         self._hostname = parsed_url.hostname
+        custom_parsed_url = urlparse(kwargs.get("custom_endpoint_address"))
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_connection.py:93]: kwargs.get("") can return None, which I think would raise an error if passed into urlparse. So these operations should probably be separated.

File: sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_connection.py [hunk 4]
```diff
@@ -90,6 +90,9 @@ def __init__(self, endpoint, **kwargs):
...
+        self.custom_hostname = custom_parsed_url.hostname
+        self.custom_port = custom_parsed_url.port
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_connection.py:95]: This assumes that a custom endpoint will always be supplied?

Comment (annatisch) [reply sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_connection.py:95]: Also - why are these being assigned to `self` when they just get passed straight into the transport?

File: sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_transport.py [hunk 5]
```diff
@@ -679,8 +681,12 @@ def connect(self):
                 http_proxy_auth = (username, password)
         try:
             from websocket import create_connection
+            if self.parsed_custom_hostname!=None:
+                url = "wss://{}:{}/$servicebus/websocket/".format(self.parsed_custom_hostname,self.parsed_custom_port)
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/azure/eventhub/_pyamqp/_transport.py:685]: This formatted URL is specific to the azure services, and so shouldn't be the pyamqp layer.
This should instead be formatted in the EventHubs/ServiceBus layer.

**Review by annatisch — COMMENTED** [review]
Looking good :)
My comments for the sync code generally apply to the async as well.
Thanks!

File: sdk/eventhub/azure-eventhub/tests/pyamqp_tests/synctests/test_websocket.py [hunk 6]
```diff
@@ -8,6 +8,7 @@
 from azure.eventhub._pyamqp import authentication, ReceiveClient
 from azure.eventhub._pyamqp.constants import TransportType
 
+@pytest.mark.skip()
```
Comment (annatisch) [inline sdk/eventhub/azure-eventhub/tests/pyamqp_tests/synctests/test_websocket.py:11]: Skipped? Why have we skipped?

**Review by annatisch — APPROVED** [review]

File: sdk/eventhub/azure-eventhub/tests/pyamqp_tests/synctests/test_websocket.py (hunk 6 above)
Comment (l0lawrence) [reply sdk/eventhub/azure-eventhub/tests/pyamqp_tests/synctests/test_websocket.py:11]: We test the websocket transport in other tests and so we decided to skip these for now as they were breaking on the pipeline. We can remove them later and maybe add an additional test to the test_send.py in the livetests for websocket transport?

**Pull request merged**